snapshot/driver.go

```diff
@@ -0,0 +1,310 @@
+}
+
+func makeTest(t *testing.T, name string, driverFn func(root string) (Driver, func(), error), fn func(t *testing.T, driver Driver, work string)) func(t *testing.T) {
+	return func(t *testing.T) {
+		// Make two directories: a driver root and a play area for the tests:
+		//
+		// 	/tmp
+		// 		work/ -> passed to test functions
+		// 		root/ -> passed to driver
+		//
+		tmpDir, err := ioutil.TempDir("", "snapshot-suite-"+name+"-")
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer os.RemoveAll(tmpDir)
+
+		root := filepath.Join(tmpDir, "root")
+		if err := os.MkdirAll(root, 0777); err != nil {
+			t.Fatal(err)
+		}
+
+		driver, cleanup, err := driverFn(root)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer cleanup()
+
+		work := filepath.Join(tmpDir, "work")
+		if err := os.MkdirAll(work, 0777); err != nil {
+			t.Fatal(err)
+		}
+
+		defer dumpDir(tmpDir)
+		fn(t, driver, work)
+	}
+}
+
+// checkDriverBasic tests the basic workflow of a snapshot driver.
+func checkDriverBasic(t *testing.T, driver Driver, work string) {
+	preparing := filepath.Join(work, "preparing")
+	if err := os.MkdirAll(preparing, 0777); err != nil {
+		t.Fatal(err)
+	}
+
+	mounts, err := driver.Prepare(preparing, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(mounts) < 1 {
+		t.Fatal("expected mounts to have entries")
+	}
+
+	if err := containerd.MountAll(mounts, preparing); err != nil {
+		t.Fatal(err)
+	}
+	defer testutil.Unmount(t, preparing)
+
+	if err := ioutil.WriteFile(filepath.Join(preparing, "foo"), []byte("foo\n"), 0777); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := os.MkdirAll(preparing+"/a/b/c", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	committed := filepath.Join(work, "committed")
+	if err := driver.Commit(committed, preparing); err != nil {
+		t.Fatal(err)
+	}
+
+	parent, err := driver.Parent(committed)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if parent != "" {
+		t.Fatalf("parent of new layer should be empty, got driver.Parent(%q) == %q", committed, parent)
+	}
+
+	next := filepath.Join(work, "nextlayer")
+	if err := os.MkdirAll(next, 0777); err != nil {
+		t.Fatal(err)
+	}
+
+	mounts, err = driver.Prepare(next, committed)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := containerd.MountAll(mounts, next); err != nil {
+		t.Fatal(err)
+	}
+	defer testutil.Unmount(t, next)
+
+	if err := ioutil.WriteFile(filepath.Join(next, "bar"), []byte("bar\n"), 0777); err != nil {
+		t.Fatal(err)
+	}
+
+	// also, change content of foo to bar
+	if err := ioutil.WriteFile(filepath.Join(next, "foo"), []byte("bar\n"), 0777); err != nil {
+		t.Fatal(err)
+	}
+
+	os.RemoveAll(next + "/a/b")
+	nextCommitted := filepath.Join(work, "committed-next")
+	if err := driver.Commit(nextCommitted, next); err != nil {
+		t.Fatal(err)
+	}
+
+	parent, err = driver.Parent(nextCommitted)
+	if parent != committed {
+		t.Fatalf("parent of new layer should be %q, got driver.Parent(%q) == %q", committed, next, parent)
+	}
+}
```

Comment:
And by that I mean, we will be following this up with more focused unit testing. I am waiting for @dmcgowan's `fstest` PR to land.